File: src/main/java/my/study/repository/design/pattern/strategy/UseDemo.java
package my.study.repository.design.pattern.strategy;

/**
 * Strategy patter: key word "strategy".
 * <p>
 * This patter is conform the OCP. When one thing have different algorithm to do it,
 * we should use this patter, use a context object to use the strategy.
 *
 * @author djh on  2018/9/12 16:12
 * @E-Mail 1544579459@qq.com
 */
public class UseDemo {
    public static void main(String[] args) {
        Context context = new Context();

        System.out.println("13公里Taxi费用:" + context.calculate(new TaxiPrice(), 13));

        System.out.println("13公里Bus费用:" + context.calculate(new BusPrice(), 13));
    }
}

File: src/main/java/my/study/repository/design/pattern/strategy/kotlin/UseDemo.kt
package my.study.repository.design.pattern.strategy.kotlin

/**
 * @author djh on  2018/10/15 19:30
 * @E-Mail 1544579459@qq.com
 */
object UseDemo {
    @JvmStatic
    fun main(args: Array<String>) {
        val context = Context()
    
        println("15公里 Bus 价格: ${context.calculate(Strategy.busPrice, 15)}")
    
        println("15公里 Taxi 价格: ${context.calculate(Strategy.taxiPrice, 15)}")
        
    }
}